node/node.go

```diff
@@ -90,6 +90,8 @@ type nodeChainReader interface {
 	GetTipSetStateRoot(tsKey types.SortedCidSet) (cid.Cid, error)
 	HeadEvents() *ps.PubSub
 	Load(context.Context) error
+	PutTipSetAndState(context.Context, *chain.TipSetAndState) error
```

Comment:
Blocking: we still want read write separation.  Node functions should absolutely not write to the chain store and the interface should reflect that.  Only the syncer should have this capability in production code.

It should be no problem to keep casting chainForTest to a read-write interface, or doing other function decomposition and automatic casting tricks.

Reply:
So, the issue is that we pass `node.ChainReader` to `RequirePutTsas` in `TestUpdateMessagePool` which needs `PutTipSetAndState`. It's super ugly, and I'd rather remove it, but I can't find any way to rewrite the test without using `node.ChainReader` because of all the complexity in `GenNode`.

Reply:
I'm in favor of removing `TestUpdateMessagePool`. It's way too complicated for a test that's essentially testing whether the node creates a message pool and updates it on new tip set. All the actual update logic is covered elsewhere. 

If you go that route, these methods should go away, `node.HeaviestTipSetHandled` should also go away.

Also, we need a test to properly test that the node subscribes things like the message pool to events like `NewHeaviestTipset`. This will probably require some refactoring, so it can be a new PR or just an issue.

Reply:
sounds good to me